pkg/controller/elasticsearch/driver/upgrade.go

```diff
@@ -7,248 +7,194 @@ package driver
-	//  we scheduled for a restart (maybe not restarted yet).
-
-	maxConcurrentUpgrades := 1
-	scheduledUpgrades := 0
-
-	// Only update 1 master node at a time, for safety and zen settings convenience.
-	// This can slow down the upgrade, but the number of master nodes should be small anyway.
-	maxMasterNodeUpgrades := 1
-	scheduledMasterNodeUpgrades := 0
-
-	for _, statefulSet := range ctx.statefulSets.ToUpdate() {
-		// Inspect each pod, starting from the highest ordinal, and decrement the partition to allow
-		// pod upgrades to go through, controlled by the StatefulSet controller.
-		for partition := sset.GetPartition(statefulSet); partition >= 0; partition-- {
-			if partition >= sset.GetReplicas(statefulSet) {
-				continue
-			}
-			if scheduledUpgrades >= maxConcurrentUpgrades {
-				return results.WithResult(defaultRequeue)
-			}
-			if label.IsMasterNodeSet(statefulSet) && scheduledMasterNodeUpgrades >= maxMasterNodeUpgrades {
-				return results.WithResult(defaultRequeue)
-			}
-
-			// Do we need to upgrade that pod?
-			podName := sset.PodName(statefulSet.Name, partition)
-			podRef := types.NamespacedName{Namespace: statefulSet.Namespace, Name: podName}
-			alreadyUpgraded, err := ctx.podUpgradeDone(ctx.client, ctx.esState, podRef, statefulSet.Status.UpdateRevision)
-			if err != nil {
-				return results.WithError(err)
-			}
-			if alreadyUpgraded {
-				continue
-			}
-
-			// An upgrade is required for that pod.
-			scheduledUpgrades++
-
-			// Is the pod upgrade already scheduled?
-			if partition == sset.GetPartition(statefulSet) {
-				continue
-			}
-
-			// Is the cluster ready for the node upgrade?
-			clusterReady, err := clusterReadyForNodeRestart(ctx.ES, ctx.esState)
-			if err != nil {
-				return results.WithError(err)
-			}
-			if !clusterReady {
-				// retry later
-				return results.WithResult(defaultRequeue)
-			}
-
-			log.Info("Preparing cluster for node restart", "namespace", ctx.ES.Namespace, "es_name", ctx.ES.Name)
-			if err := prepareClusterForNodeRestart(ctx.esClient, ctx.esState); err != nil {
-				return results.WithError(err)
-			}
-
-			if label.IsMasterNodeSet(statefulSet) {
-				scheduledMasterNodeUpgrades++
-				// TODO if the node is a master:
-				//  - zen1: update minimum_master_node to account for master node deletion. Otherwise upgrading a 2-masters
-				//   cluster provokes downtime since m_m_n=2.
-				//   Problem: how to prevent this to be reverted at the next reconciliation, before the pod gets deleted?
-				//  - zen2: set voting config exclusions: same problem, this is not easy. But since we only delete
-				//   one master at a time, maybe it's not required?
-			}
-
-			// Upgrade the pod.
-			if err := ctx.upgrader(&statefulSet, partition); err != nil {
-				return results.WithError(err)
-			}
-		}
+func (ctx rollingUpgradeCtx) run() ([]corev1.Pod, error) {
+	deletedPods, err := ctx.Delete()
+	if errors.IsConflict(err) || errors.IsNotFound(err) {
```

Comment:
It is a little bit tricky to move this one because I need the error to cancel the expectation, stop the deletion process and I want to hide this error _(or just print a debug message)_ because there is nothing the user can do about that.
What I can do is add more details in the comment to explain why it is happening.

Reply:
Oh, got it. I'm fine with the way it is.